pandas/tests/indexing/multiindex/test_multiindex.py

```diff
@@ -98,3 +100,23 @@ def test_multiindex_with_datatime_level_preserves_freq(self):
         result = df.loc[0].index
         tm.assert_index_equal(result, dti)
         assert result.freq == dti.freq
+
+    def test_concat_with_various_multiindex_dtypes(self):
+        # GH #23478
+        columns_list = [
+            MultiIndex.from_product([["a"], range(2)]),
+            MultiIndex.from_product([["a"], np.arange(2.0)]),
+            MultiIndex.from_product([["b"], ["A", "B"]]),
+            MultiIndex.from_product(
+                [["c"], pd.date_range(start="2017", end="2018", periods=2)]
+            ),
+        ]
+
+        dfs = [
+            DataFrame(np.zeros((1, len(columns))), columns=columns)
+            for columns in columns_list
+        ]
+
+        for i, j in combinations(range(len(dfs)), 2):
```

Comment:
can you parameterize instead of doing thsi.

pls check the warning text as well.

Reply:
Thanks for the comment. I have attempted to parameterize the test as requested. I suspect I am not using ```assert_produces_warning``` correctly, could you give any hints how I should correctly check for the TypeErrors?